google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,234 @@
+  public void setSdkHomePath(String path) {
+    /* unsupported, to be removed. */
+  }
 
   @Override
   public SdkStatus getStatus() {
-    return SdkStatus.NOT_AVAILABLE;
+    return sdkStatus;
   }
 
   @Override
-  public boolean installAutomatically() {
+  public boolean install() {
+    // check if installation is already running first, to prevent multiple jobs running.
+    if (runningInstallationJob == null || runningInstallationJob.isDone()) {
+      runningInstallationJob =
+          ThreadUtil.getInstance().executeInBackground(this::installManagedSdk);
+      Futures.addCallback(
+          runningInstallationJob,
+          new FutureCallback<Path>() {
+            @Override
+            public void onSuccess(Path path) {
+              if (path != null) {
+                logger.info("Managed Google Cloud SDK successfully installed at: " + path);
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              if (t instanceof InterruptedException) {
+                logger.info("Managed Google Cloud SDK installation cancelled.");
+              }
+            }
+          });
+    }
+
     return true;
   }
 
   @Override
-  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {}
+  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.add(listener);
+  }
+
+  @Override
+  public void removeStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.remove(listener);
+  }
+
+  public void cancelInstall() {
+    if (runningInstallationJob != null) {
+      runningInstallationJob.cancel(true);
+    }
+  }
+
+  @VisibleForTesting
+  ManagedCloudSdk createManagedSdk() throws UnsupportedOsException {
+    return ManagedCloudSdk.newManagedSdk();
+  }
+
+  @VisibleForTesting
+  void setLogger(Logger logger) {
+    this.logger = logger;
+  }
+
+  @VisibleForTesting
+  void invokeOnApplicationUIThread(Runnable runnable) {
+    ApplicationManager.getApplication().invokeLater(runnable);
+  }
+
+  /**
+   * Checks for Managed Cloud SDK status and creates/installs it if necessary.
+   *
+   * @return Path of the installed managed SDK, null if error occurred while installing.
+   * @throws InterruptedException if Managed Cloud SDK has been interrupted by {@link
+   *     #cancelInstall()}
+   */
+  private Path installManagedSdk() throws InterruptedException {
+    if (managedCloudSdk == null) {
+      try {
+        managedCloudSdk = createManagedSdk();
+      } catch (UnsupportedOsException ex) {
+        logger.error("Unsupported OS for Managed Cloud SDK", ex);
+        updateStatus(SdkStatus.NOT_AVAILABLE);
+        return null;
+      }
+    }
+
+    updateStatus(SdkStatus.INSTALLING);
+
+    Path installedManagedSdkPath;
+    try {
+      installedManagedSdkPath = installSdk();
+    } catch (InterruptedException iex) {
+      updateStatus(SdkStatus.NOT_AVAILABLE);
+      throw iex;
```

Comment:
follow up to my other comment: it seems a little bit inconsistent that you are throwing this one, but swallowing the other ones and returning null instead.

Reply:
It does feel filled with various exceptions, I discussed with Appu which ones means what and it's not pretty.
There are a few things here why this happens:
* InterruptedException is not an error but sign that process was gracefully cancelled so that's the reason it goes out to the top and will be logged in UI
* ManagedSdkVerificationException | ManagedSdkVersionMismatchException are not fatal and means SDK was there but it failed to execute status check on them - it just means we go ahead and re-install it. Those happen for both base SDK and App Engine Java and we need to log different messages about them.
* Other exceptions are fatal.

So it's possible `onFailure()` can become really ugly method, especially when we need to distinguish between SDK / App Engine Java component errors. WDYT?

Reply:
hmm thinking about this

Reply:
> InterruptedException is not an error but sign that process was gracefully cancelled so that's the reason it goes out to the top and will be logged in UI

why not just do the log.info in the same place rather then letting the exception propagate out then? I just think there is a slight readability / consistency issue here

Reply:
ah I think I see part of your dilemma. you need to handle equivalent exceptions differently if its installing the sdk vs installing the java component

Reply:
Yes, that was the main reason - same exceptions for two different processes (we can combine them but then can loose exact understanding of what happened.)
As for handling `InterruptedException` it's handled on top level since I expect to call UI (notification bubble) for it, so I didn't want to call UI notification code from `install()` method family. I guess that would be cleaner with UI updates in place but that could have made PR too big.